### ISSUE TITLE
Some `staticmethod`s with hardcoded classname make inheritance ugly

### ISSUE THREAD

**Issue by mahnerak** [allenai/allennlp#3196]
In the framework in multiple places methods like `from_files`, `from_params` etc. are implemented as `staticmethod`s. Note, that these methods almost always return a new instance of the same type. For example:
```python
class Params(MutableMapping):
    ...
    @staticmethod
    def from_file(params_file: str, params_overrides: str = "", ext_vars: dict = None) -> 'Params':
        ....
        return Params(param_dict)
```
In case of if you want to inherit from `Param` a new class `SubclassOfParam` to alter behaviour of certain methods (say add one more config serialization step in the method `to_file`), I have to copy-paste contents of `Params.from_file()` to make sure `SubclassOfParam.from_file()` returns instance of type `SubclassOfParam`, not `Param`.

Are there any reasons to not to refactor all of these methods to make them `classmethod`s instead of `staticmethod`s?

P.S. Previously we had a conversation on `Vocabulary`: #2418

**Comment by mahnerak** [comment]
We had similar discussion on `Vocabulary` class: #2418

**Comment by DeNeutoy** [comment]
Fixed by #3229
